Source/PeleLM.cpp

```diff
@@ -7616,39 +7626,57 @@ PeleLM::compute_Wbar_fluxes(Real time,
-      crse_br.copy<RunOn::Host>From(Wbar_crse,nGrowCrse,0,0,1);
-      bndry->setBndryValues(crse_br,0,Wbar,0,0,1,crse_ratio,bc);
+      visc_op.setCoarseFineBC(&Wbar_crse, crse_ratio[0]);
+   }
+   visc_op.setLevelBC(0, &Wbar);
+
+   visc_op.setScalars(0.0,1.0);
+   {
+      FluxBoxes fb_Coeff(this,1,0);
+      MultiFab** bcoeffs = fb_Coeff.get();
+      for (int d=0; d<AMREX_SPACEDIM; ++d) {
+         bcoeffs[d]->setVal(1.0);
+      }
+      std::array<MultiFab*,AMREX_SPACEDIM> fp{AMREX_D_DECL(bcoeffs[0],bcoeffs[1],bcoeffs[2])};
+      visc_op.setBCoeffs(0, amrex::GetArrOfConstPtrs(fp));
    }
 
-   visc_op->setScalars(0,1);
-   visc_op->bCoefficients(1);
-   visc_op->applyBC(Wbar,0,1,0,LinOp::Inhomogeneous_BC);
+   MLMG mg(visc_op);   
 
-   delete visc_op;
+   FluxBoxes fb_flux(this,1,0);
+   MultiFab** gradWbar = fb_flux.get();
+
+   std::array<MultiFab*,AMREX_SPACEDIM> fp{AMREX_D_DECL(gradWbar[0],gradWbar[1],gradWbar[2])};
+   mg.getFluxes({fp},{&Wbar},MLLinOp::Location::FaceCentroid);
 
 #ifdef _OPENMP
-#pragma omp parallel
+#pragma omp parallel if (Gpu::notInLaunchRegion())
 #endif   
-   for (MFIter mfi(Wbar,true); mfi.isValid(); ++mfi)
+   for (MFIter mfi(Wbar,TilingIfNotGPU()); mfi.isValid(); ++mfi)
    {
-     
-      const FArrayBox& wbar = Wbar[mfi];
-      const Real       mult = -1.0;
-
       for (int d=0; d<AMREX_SPACEDIM; ++d) 
       {
-         const Box&        vbox = mfi.nodaltilebox(d);
-         const FArrayBox& rhoDe = (*betaWbar[d])[mfi];
-         FArrayBox&    fluxfab  = (*SpecDiffusionFluxWbar[d])[mfi];
+         const Box&  vbx = mfi.nodaltilebox(d);
+         auto const& gradWbar_ar  = gradWbar[d]->array(mfi);
+         auto const& betaWbar_ar  = betaWbar[d]->array(mfi);
+         auto const& wbarFlux     = SpecDiffusionFluxWbar[d]->array(mfi);
 
-         for (int ispec=0; ispec<nspecies; ++ispec)
-         {
-           grad_wbar(BL_TO_FORTRAN_BOX(vbox),
-                     BL_TO_FORTRAN_ANYD(wbar),
-                     BL_TO_FORTRAN_ANYD(rhoDe),
-                     BL_TO_FORTRAN_N_ANYD(fluxfab,ispec),
-                     BL_TO_FORTRAN_ANYD(area[d][mfi]),
-                     &dx[d], &d, &mult, &inc);
+         if ( increment_flag == 0 ) {                 // Overwrite wbar fluxes
+            amrex::ParallelFor(vbx, [gradWbar_ar, betaWbar_ar, wbarFlux]
+            AMREX_GPU_DEVICE (int i, int j, int k) noexcept
+            {
+               for (int n = 0; n <= NUM_SPECIES; n++) {
+                  wbarFlux(i,j,k,n) = - betaWbar_ar(i,j,k,n) * gradWbar_ar(i,j,k);
+               }   
+            });
+         } else {                                     // Increment wbar fluxes
+            amrex::ParallelFor(vbx, [gradWbar_ar, betaWbar_ar, wbarFlux]
+            AMREX_GPU_DEVICE (int i, int j, int k) noexcept
+            {
+               for (int n = 0; n <= NUM_SPECIES; n++) {
+                  wbarFlux(i,j,k,n) -= betaWbar_ar(i,j,k,n) * gradWbar_ar(i,j,k);
+               }   
+            });
```

Comment:
I just didn't want to call the getFluxes NUM_SPECIES times when all I really need is gradWbar. I thought it might be faster to go this way, but maybe I'm wrong.

Reply:
Nah, you're right.  I wasn't seeing the entire thing clearly enough.